### ISSUE TITLE
Is there a point in fine tuning from checkpoint if classes differ?

### ISSUE THREAD

**Issue by mosheliv** [open-mmlab/mmdetection#1116]
Supposed I have a dataset with 200 classes. What would make more sense, training from scratch or fine tuning from the modelzoo coc checkpoint using https://github.com/open-mmlab/mmdetection/issues/743 ?
In classification, its usually better to fine tune but in object detection... I don't know. I guess I could try both but it takes so long on single gpu...

**Comment by hellock** [comment]
You can try the finetuning.

**Comment by mosheliv** [comment]
I know I can, I was just asking (before running training for a few days
finding its worse) if you have any experience with fine tuning working
faster then starting from scratch.

**Comment by hellock** [comment]
There is no exact conclusion and it depends on your dataset. Training from scratch requires a large dataset and takes 3x to 6x longer time than finetuning from ImageNet pretrained models or COCO pretrained models. [Here](https://github.com/open-mmlab/mmdetection/blob/master/configs/scratch/README.md) are some reference.

**Comment by mosheliv** [comment]
there is a little ambiguity when it comes to the "fine tuning" in object
detection. The reference you gave referred to using pre trained feature
extractors (classifier backbone) trained on imagenet, which is now more or
less the most common practice.
In object detectors (particularly two stage), there is another decision to
make.
suppose you have a detector trained on coco and you want it to detect on a
subset of coco (like person and car), it would be much much faster to fine
tune the model, i.e. take the coco model, remove the last layers, add new
ones that has the correct number of classes and retrain, probably with a
lower learning rate and less epochs as basically the detector is well aware
of these classes.
In my case, it is slightly different as I have MORE classes then coco and
more diverse although there is some overlap.

My question is, do you have an experience with  doing fine tuning (in the
object detection sense) for such dataset? do you have any ballpark
recommendations for the lr schedule?

**Comment by mosheliv** [comment]
I have decided to go ahead and try it with retinanet. I used:
```
import torch
import pprint

pretrained_weights  = torch.load('/lh/retinanet_x101_64x4d_fpn_1x_20181218-a0a22662.pth')
#pprint.pprint(pretrained_weights['state_dict'])

num_class = 101
pretrained_weights['state_dict']['bbox_head.retina_cls.weight'].resize_((num_class-1)*9, 256, 3, 3)
pretrained_weights['state_dict']['bbox_head.retina_cls.bias'].resize_((num_class-1)*9)
#pretrained_weights['state_dict']['bbox_head.retina_reg.weight'].resize_(num_class*4, 1024)
#pretrained_weights['state_dict']['bbox_head.retina_reg.bias'].resize_(num_class*4)

torch.save(pretrained_weights, "/lh/retinanet_x101_64x4d_fpn_1x_ft.pth")

```
to convert the model but got inf loss 
my config is pretty much the standard one with load_from the provided checkpoint in the zoo with 101 classes.
Any ideas what am I doing wrong?

**Comment by mosheliv** [comment]
an example of a training session:
```
2019-08-11 08:41:16,546 - INFO - load checkpoint from /lh/retinanet_x101_64x4d_fpn_1x_ft.pth
2019-08-11 08:41:16,800 - INFO - Start running, host: root@133baf65d100, work_dir: /lh/e7_ft
2019-08-11 08:41:16,801 - INFO - workflow: [('train', 1)], max: 12 epochs
2019-08-11 08:42:44,726 - INFO - Epoch [1][50/46569]    lr: 0.00050, eta: 11 days, 8:55:49, time: 1.758, data_time: 0.014, memory: 9489, loss_cls: inf, loss_b
box: 149046879565130366976.0000, loss: inf
2019-08-11 08:44:15,292 - INFO - Epoch [1][100/46569]   lr: 0.00058, eta: 11 days, 13:00:48, time: 1.811, data_time: 0.005, memory: 9489, loss_cls: inf, loss_
bbox: 148944405218339657351168.0000, loss: inf

```

**Comment by mosheliv** [comment]
actually I think I know what the problem is. This was training on my secondary gpu which can only run 2 images in a batch and I seem to recall this caused similar behavior before. I'll wait for my primary gpu to clear...

**Comment by DecentMakeover** [comment]
@mosheliv What command did you run to load from checkpoint, when i run this  python3 tools/train.py configs/fast_rcnn_r50_fpn_1x.py --load_from checkpoints/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth 

train.py: error: unrecognized arguments: --load_from checkpoints/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth

**Comment by mosheliv** [comment]
I added the load from to the train script... Just copy the resume from
lines and change

**Comment by DecentMakeover** [comment]
@mosheliv  Got it, Thanks.

**Comment by mosheliv** [comment]
Please let me know what were your results... My gpu will clear in 4 days...

**Comment by DecentMakeover** [comment]
Even i am not sure when i will run the experiments , i am just doing a sanity check on the code.Once i do run , i will surely post.

**Comment by DecentMakeover** [comment]
@mosheliv  hey did you run?

**Comment by mosheliv** [comment]
Not yet... My gpus where very busy. Perhaps this week...
